### PR TITLE
Aes128gcm support

### PR DESCRIPTION
**Pull request by pinarx**
Add support for aes128gcm changes.

### CODE REVIEW

**Comment by pinarx** [comment]
signed

File: webpush/encrypt.go [hunk 1]
```diff
@@ -55,12 +55,36 @@ import (
 	"strings"
 )
 
+// Encryption version.
+type Version int
+
+const (
+	// The version that is widely deployed with WebPush (as of 2016-11).
+	AESGCM = iota
```
Comment (gauntface) [inline webpush/encrypt.go]: What is 'iota' here?

File: webpush/encrypt.go [hunk 2]
```diff
@@ -55,12 +55,36 @@ import (
...
+)
+
+func (v Version) String() string {
+	switch v {
+	case AESGCM:
+		return "aesgcm"
+	case AES128GCM:
+		return "aes128gcm"
+	}
+	return ""
```
Comment (gauntface) [inline webpush/encrypt.go]: Is there ever a scenario with an empty string will actually be returned? My naivety to Go says this should thrown error rather than return an empty string.

File: webpush/encrypt.go [hunk 3]
```diff
@@ -144,16 +168,25 @@ type EncryptionResult struct {
...
+	if version == AESGCM {
+		maxPayloadLength = aesgcmMaxPayloadLength
+	} else {
+		maxPayloadLength = aes128gcmMaxPayloadLength
+	}
 	if len(plaintext) > maxPayloadLength {
 		return nil, fmt.Errorf("Payload is too large. The max number of bytes is %d, input is %d bytes.", maxPayloadLength, len(plaintext))
 	}
 
+	// sub.Key is the user agent's public key.
```
Comment (gauntface) [inline webpush/encrypt.go]: This is the subscriptions public key rather then than the user agents key right? Also the naming makes this a little hard to parse, i.e. I'm assuming sub is subscription and Key is too vague to be meaningful in the context of an encryption flow. p256dh would be at least more obvious.

File: webpush/encrypt.go (hunk 1 above)
Comment (anaskhan96) [reply webpush/encrypt.go]: I think as far as I know, `iota` is a way to declare numeric constants from a starting value (default 0) which increments by 1. So in this case, `AESGCM` = 0, and `AES128GCM` = 1.

File: webpush/encrypt.go [hunk 4]
```diff
@@ -241,11 +292,42 @@ func newInfo(infoType string, context []byte) []byte {
...
+	if len(context) > 0 {
+		info = append(info, []byte("P-256")...)
+		info = append(info, context...)
+	}
 	return info
 }
 
+// Returns a keying material. See sections of
+// https://tools.ietf.org/html/draft-ietf-webpush-encryption-07#section-3.3
+func newKeyInfo(userAgentPublicKey, serverPublicKey []byte) []byte {
```
Comment (gauntface) [inline webpush/encrypt.go]: Possible to rename userAgentPublicKey to subscriptionPublicKey or p256dhKey ?

File: webpush/encrypt.go [hunk 5]
```diff
@@ -55,12 +55,36 @@ import (
...
 const (
-	maxPayloadLength = 4078
+	aesgcmMaxPayloadLength = 4078
+	// due to binary message header.
+	aes128gcmMaxPayloadLength = 4057
 )
 
 var (
 	authInfo = []byte("Content-Encoding: auth\x00")
+	keyInfoPrefix = []byte("WebPush: info\x00")
```
Comment (gauntface) [inline webpush/encrypt.go]: I can't find a reference to this header anywhere, do you have a link you could include in a comment?

File: webpush/encrypt.go [hunk 6]
```diff
@@ -55,12 +55,36 @@ import (
...
+
+const (
+	// The version that is widely deployed with WebPush (as of 2016-11).
+	AESGCM = iota
+	// The most recent version, the salt, record size and key identifier
+	// are included in a header that is part of the encrypted content coding.
+	AES128GCM
+)
+
+func (v Version) String() string {
```
Comment (gauntface) [inline webpush/encrypt.go]: Worth changing Version to ContentEncoding to be more explicit?

File: webpush/encrypt.go [hunk 7]
```diff
@@ -241,11 +292,42 @@ func newInfo(infoType string, context []byte) []byte {
...
+	return info
+}
+
+// Returns an 86 octet header. See section 2.1 of:
+// https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-06#section-2.1
+func appendHeader(salt, serverPublicKey, ciphertext []byte) []byte {
+	var result []byte
+	// A push service is not required to support more than 4096 octets of
+	// payload body so the record size can be at most 4096.
+	cplen := uint32(4096)
```
Comment (gauntface) [inline webpush/encrypt.go]: This is a push service minimum requirement, not a maximum, should it be enforced?

Comment (gauntface) [reply webpush/encrypt.go]: Also not sure what 'cplen' is....

File: webpush/encrypt.go [hunk 8]
```diff
@@ -276,14 +358,18 @@ func hkdf(salt, ikm, info []byte, length int) []byte {
...
-	// padding.
-	// TODO: Right now we leave the size at zero. We should add a padding option
-	// that allows the payload size to be obscured.
-	padding := make([]byte, 2)
-	data := append(padding, plaintext...)
-
+func encrypt(plaintext, key, nonce []byte, version Version) ([]byte, error) {
+	var data []byte
+	if version == AES128GCM {
+		data = append(plaintext, []byte{0x02}...)
```
Comment (gauntface) [inline webpush/encrypt.go]: What is this for? Looks like still adding padding but at the end instead of the start but I'm not sure if I'm reading the syntax correct (i.e. what is the 0x02 hex value mean in this?

Comment (gauntface) [reply webpush/encrypt.go]: Also is padding still possible in the spec?

File: webpush/encrypt.go (hunk 1 above)
Comment (pinarx) [reply webpush/encrypt.go]: See: https://github.com/golang/go/wiki/Iota

File: webpush/encrypt.go (hunk 2 above)
Comment (pinarx) [reply webpush/encrypt.go]: good point, changing into panic.

File: webpush/encrypt.go (hunk 3 above)
Comment (pinarx) [reply webpush/encrypt.go]: yes but it's issued by the user agent, thus it's clearer to me to associate it with the user agent. But agreed that p256dh is more explicit.

File: webpush/encrypt.go (hunk 4 above)
Comment (pinarx) [reply webpush/encrypt.go]: chose subscriptionPublicKey to clarify it is retrieved from the PushSubscription

**Review by gauntface — CHANGES_REQUESTED** [review]
Sorry I didn't post these extra comments :(

File: webpush/encrypt.go (hunk 6 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 5 above)
Comment (pinarx) [reply webpush/encrypt.go]: It's in here: https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#section-3.3

Added as a comment.

File: webpush/encrypt.go (hunk 7 above)
Comment (pinarx) [reply webpush/encrypt.go]: The intermediary values in the spec seems to enforce, that's why so did I: https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#appendix-A

And I believe the RFC tests also depend on it, since that's the way it's used in the spec.

For cplen: I think I just copied from the existing version but it's referring to client public key length probably=) Will change it to plen for payload length

File: webpush/encrypt.go (hunk 8 above)
Comment (pinarx) [reply webpush/encrypt.go]: "The padding delimiter octet MUST be checked, values other than 0x02 MUST cause the message to be discarded."

From: https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#section-4

Added a comment there as well

**Review by gauntface — APPROVED** [review]

File: webpush/encrypt.go [hunk 9]
```diff
@@ -55,12 +55,38 @@ import (
 	"strings"
 )
 
+// Encryption encoding.
+type ContentEncoding int
```
Comment (x1ddos) [inline webpush/encrypt.go]: Exported type doc comments should start with the symbol: "ContentEncoding ..."
Here and everywhere else.

File: webpush/encrypt.go [hunk 10]
```diff
@@ -55,12 +55,38 @@ import (
...
+	case AES128GCM:
+		return "aes128gcm"
+	}
+	panic("A valid content encoding must be used")
+}
+
 const (
-	maxPayloadLength = 4078
+	aesgcmMaxPayloadLength = 4078
+	// due to binary message header.
```
Comment (x1ddos) [inline webpush/encrypt.go]: Start with a capital letter. Also, not quite of a useful comment: is the binary header too big, too small or are there other restrictions? Otherwise, you might as well remove the comment.

File: webpush/encrypt.go [hunk 11]
```diff
@@ -55,12 +55,38 @@ import (
 	"strings"
 )
 
+// Encryption encoding.
+type ContentEncoding int
+
+const (
+	// The encoding that is widely deployed with WebPush (as of 2016-11).
+	AESGCM = iota
```
Comment (x1ddos) [inline webpush/encrypt.go]: Specify the type:

```go
AESGCM ContentEncoding = iota
...
```

See https://github.com/golang/go/wiki/Iota.

File: webpush/encrypt.go [hunk 12]
```diff
@@ -144,16 +170,25 @@ type EncryptionResult struct {
 //    - https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding
 //    - https://en.wikipedia.org/wiki/Elliptic_curve_Diffie%E2%80%93Hellman
 //    - https://tools.ietf.org/html/draft-ietf-webpush-encryption
-func Encrypt(sub *Subscription, message string) (*EncryptionResult, error) {
+func Encrypt(sub *Subscription, message string, encoding ContentEncoding) (*EncryptionResult, error) {
 	plaintext := []byte(message)
+
+	var maxPayloadLength int
```
Comment (x1ddos) [inline webpush/encrypt.go]: There's no reason to define a var with initial zero value here. A more Go idiomatic style would be the following:

```go
maxlen = aesgcmMaxPayloadLength
if encoding == AES128GCM {
  maxlen = aes128gcmMaxPayloadLength
}
```

File: webpush/encrypt.go [hunk 13]
```diff
@@ -144,16 +170,25 @@ type EncryptionResult struct {
...
+		maxPayloadLength = aesgcmMaxPayloadLength
+	} else {
+		maxPayloadLength = aes128gcmMaxPayloadLength
+	}
 	if len(plaintext) > maxPayloadLength {
 		return nil, fmt.Errorf("Payload is too large. The max number of bytes is %d, input is %d bytes.", maxPayloadLength, len(plaintext))
 	}
 
+	// sub.Key is the p256dh key.
 	if len(sub.Key) == 0 {
```
Comment (x1ddos) [inline webpush/encrypt.go]: You could check this and sub.Auth at the top of the function, before anything else, to avoid unnecessary allocations. It would also make the main flow more readable

File: webpush/encrypt.go [hunk 14]
```diff
@@ -241,11 +294,42 @@ func newInfo(infoType string, context []byte) []byte {
...
+	return info
+}
+
+// Returns an 86 octet header. See section 2.1 of:
+// https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-06#section-2.1
+func appendHeader(salt, serverPublicKey, ciphertext []byte) []byte {
+	var result []byte
+	// A push service is not required to support more than 4096 octets of
+	// payload body so the record size can be at most 4096.
+	plen := uint32(4096)
```
Comment (x1ddos) [inline webpush/encrypt.go]: Might as well be a constant, defined at the top. Seems important enough.

File: webpush/encrypt.go [hunk 15]
```diff
@@ -241,11 +294,42 @@ func newInfo(infoType string, context []byte) []byte {
...
 }
 
+// Returns a keying material. See sections of
+// https://tools.ietf.org/html/draft-ietf-webpush-encryption-07#section-3.3
+func newKeyInfo(subscriptionPublicKey, serverPublicKey []byte) []byte {
+	var info []byte
+	info = append(info, keyInfoPrefix...)
+	info = append(info, subscriptionPublicKey...)
+	info = append(info, serverPublicKey...)
+	return info
```
Comment (x1ddos) [inline webpush/encrypt.go]: Because you know all the lengths, you could also make it a bit more performant:

```go
info := make([]byte, len(keyInfoPrefix) + len(subscriptionPublicKey) + len(serverPublicKey))
n := copy(info, keyInfoPrefix)
n += copy(info[n:], subscriptionPublicKey)
copy(info[n:], serverPublicKey)
```

File: webpush/encrypt.go [hunk 16]
```diff
@@ -241,11 +294,42 @@ func newInfo(infoType string, context []byte) []byte {
...
+		info = append(info, context...)
+	}
 	return info
 }
 
+// Returns a keying material. See sections of
+// https://tools.ietf.org/html/draft-ietf-webpush-encryption-07#section-3.3
+func newKeyInfo(subscriptionPublicKey, serverPublicKey []byte) []byte {
+	var info []byte
+	info = append(info, keyInfoPrefix...)
```
Comment (x1ddos) [inline webpush/encrypt.go]: or just append to a literal: `info := append([]byte{}, keyInfoPrefix...)`

File: webpush/encrypt.go [hunk 17]
```diff
@@ -241,11 +294,42 @@ func newInfo(infoType string, context []byte) []byte {
...
+	var result []byte
+	// A push service is not required to support more than 4096 octets of
+	// payload body so the record size can be at most 4096.
+	plen := uint32(4096)
+	plenbuf := make([]byte, 5)
+	binary.BigEndian.PutUint32(plenbuf, plen)
+	binary.BigEndian.PutUint16(plenbuf[3:], uint16(len(serverPublicKey)))
+	result = append(result, salt...)
+	result = append(result, plenbuf...)
+	result = append(result, serverPublicKey...)
```
Comment (x1ddos) [inline webpush/encrypt.go]: This doesn't seem to match what is described in https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-06#section-2.1 or am I missing something?

File: webpush/encrypt.go [hunk 18]
```diff
@@ -276,14 +360,22 @@ func hkdf(salt, ikm, info []byte, length int) []byte {
...
-	// TODO: Right now we leave the size at zero. We should add a padding option
-	// that allows the payload size to be obscured.
-	padding := make([]byte, 2)
-	data := append(padding, plaintext...)
-
+func encrypt(plaintext, key, nonce []byte, encoding ContentEncoding) ([]byte, error) {
+	var data []byte
+	if encoding == AES128GCM {
+		// The padding delimiter octet MUST be checked, values other than 0x02 MUST
+    // cause the message to be discarded.
```
Comment (x1ddos) [inline webpush/encrypt.go]: Misindented comment line.

File: webpush/encrypt.go [hunk 19]
```diff
@@ -276,14 +360,22 @@ func hkdf(salt, ikm, info []byte, length int) []byte {
...
+		// See here for more details for restrictions on use of "aes128gcm":
+		// https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#section-4
+		data = append(plaintext, []byte{0x02}...)
+	} else {
+		// Add padding. There is a uint16 size followed by that number of bytes of
+		// padding.
+		// TODO: Right now we leave the size at zero. We should add a padding option
+		// that allows the payload size to be obscured.
+		padding := make([]byte, 2)
+		data = append(padding, plaintext...)
```
Comment (x1ddos) [inline webpush/encrypt.go]: no need for `make`, just do append to a literal: `append([]byte{0,0}, plaintext...)`

File: webpush/encrypt_test.go [hunk 20]
```diff
@@ -36,9 +36,15 @@ var (
...
-	rfcAuth   = "R29vIGdvbyBnJyBqb29iIQ"
+	aes128gcmMessage   = `When I grow up, I want to be a watermelon`
+
+	rfcAesgcmPublic = "BCEkBjzL8Z3C-oi2Q7oE5t2Np-p7osjGLg93qUP0wvqRT21EEWyf0cQDQcakQMqz4hQKYOQ3il2nNZct4HgAUQU"
+	rfcAes128gcmPublic = "BCVxsr7N_eNgVRqvHtD0zTZsEc6-VV-JvLexhqUzORcxaOzi6-AYWXvTBHm4bjyPjs7Vd8pZGH6SRpkNtoIAiw4"
+
+	rfcAesgcmCipher = "6nqAQUME8hNqw5J3kl8cpVVJylXKYqZOeseZG8UueKpA"
+	rfcAes128gcmCipher = "DGv6ra1nlYgDCS1FRnbzlwAAEABBBP4z9KsN6nGRTbVYI_c7VJSPQTBtkgcy27mlmlMoZIIgDll6e3vCYLocInmYWAmS6TlzAC8wEqKK6PBru3jl7A_yl95bQpu6cVPTpK4Mqgkf1CXztLVBSt2Ks3oZwbuwXPXLWyouBWLVWGNWQexSgSxsj_Qulcy4a-fN"
+	rfcAesgcmAuth   = "R29vIGdvbyBnJyBqb29iIQ"
+	rfcAes128gcmAuth   = "BTBZMqHH6r4Tts7J_aSIgg"
```
Comment (x1ddos) [inline webpush/encrypt_test.go]: I don't think the above vars are properly formatted. Run `fmt` on this file.

File: webpush/encrypt_test.go [hunk 21]
```diff
@@ -169,31 +185,44 @@ func rfcKeys() ([]byte, []byte, error) {
 	return priv, elliptic.Marshal(curve, x, y), nil
 }
 
-// TestRfcVectors uses the values given in the RFC for HTTP encryption to verify
+func rfcAes128gcmSalt() ([]byte, error) {
```
Comment (x1ddos) [inline webpush/encrypt_test.go]: I'd say `AES` is an initialism: https://github.com/golang/go/wiki/CodeReviewComments#initialisms
Here and everywhere else.

File: webpush/encrypt_test.go [hunk 22]
```diff
@@ -202,6 +231,39 @@ func TestRfcVectors(t *testing.T) {
 	}
 }
 
+// TestAes128gcmRfcVectors uses the values given in the RFC for HTTP encryption to verify
+// that the code conforms to the RFC
+// See: https://tools.ietf.org/html/draft-ietf-webpush-encryption-07#appendix-A
+func TestAes128gcmRfcVectors(t *testing.T) {
```
Comment (x1ddos) [inline webpush/encrypt_test.go]: It actually makes it harder to read the function code with so many empty lines between expressions.

File: webpush/encrypt_test.go [hunk 23]
```diff
@@ -202,6 +231,39 @@ func TestRfcVectors(t *testing.T) {
...
+	defer stubFuncs(rfcAes128gcmSalt, rfcAes128gcmKeys)()
+
+	b64 := base64.URLEncoding.WithPadding(base64.NoPadding)
+
+	auth, err := b64.DecodeString(rfcAes128gcmAuth)
+	if err != nil {
+		t.Error(err)
```
Comment (x1ddos) [inline webpush/encrypt_test.go]: `t.Fatal(err)` - you can't use DecodeString results at all if an error was encountered.

**Review by x1ddos — CHANGES_REQUESTED** [review]

File: webpush/encrypt.go (hunk 9 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 11 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 10 above)
Comment (pinarx) [reply webpush/encrypt.go]: Added more detail.

File: webpush/encrypt.go (hunk 12 above)
Comment (pinarx) [reply webpush/encrypt.go]: defaulted to the aes128

File: webpush/encrypt.go (hunk 15 above)
Comment (pinarx) [reply webpush/encrypt.go]: cool, thanks

File: webpush/encrypt.go (hunk 16 above)
Comment (pinarx) [reply webpush/encrypt.go]: took the alternative approach you suggested

File: webpush/encrypt.go (hunk 14 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 18 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 19 above)
Comment (pinarx) [reply webpush/encrypt.go]: This was from the previous implementation - but done

File: webpush/encrypt_test.go (hunk 21 above)
Comment (pinarx) [reply webpush/encrypt_test.go]: done

File: webpush/encrypt_test.go (hunk 22 above)
Comment (pinarx) [reply webpush/encrypt_test.go]: got rid of some?

File: webpush/encrypt_test.go (hunk 23 above)
Comment (pinarx) [reply webpush/encrypt_test.go]: ah good catch

File: webpush/encrypt.go (hunk 17 above)
Comment (pinarx) [reply webpush/encrypt.go]: Which part?
Also see: https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#section-3.1 and https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#appendix-A

File: webpush/encrypt_test.go (hunk 20 above)
Comment (pinarx) [reply webpush/encrypt_test.go]: done

File: webpush/encrypt.go (hunk 13 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 17 above)
Comment (beverloo) [reply webpush/encrypt.go]: @x1ddos is right -- the `idlen` field in the binary header is a uint8 value, but it's being written on line 326 as a uint16 value. Writing it as such will yield [0x00, 0x41], of which the user agent will interpret the first byte (0x00) as the length of the public key. That's not right.

In general, `rs` (uint32) and `idlen` (uint8) are two separate values, I would treat them as such and not put them together in the 5-byte `plenbuf` field.

Comment (pinarx) [reply webpush/encrypt.go]: got it, fixed the buffers.

File: webpush/encrypt.go [hunk 24]
```diff
@@ -55,13 +55,43 @@ import (
...
+)
+
+func (v ContentEncoding) String() string {
+	switch v {
+	case AESGCM:
+		return "aesgcm"
+	case AES128GCM:
+		return "aes128gcm"
+	}
+	panic("A valid content encoding must be used")
```
Comment (x1ddos) [inline webpush/encrypt.go]: Go discourages panic usage, especially in packages imported by others. You could just return something like "unknown".
if you want to ensure a valid value, you'd typically do it in a function which takes it as an argument. For example, `Encrypt`, in which case you wouldn't panic anyway and return an error instead.

Either way, this is definitely not the place for panic.

File: webpush/encrypt.go [hunk 25]
```diff
@@ -144,20 +174,27 @@ type EncryptionResult struct {
...
 		return nil, fmt.Errorf("Subscription must include the client's public key")
 	}
 
+	// sub.Auth is the authentication secret.
 	if len(sub.Auth) == 0 {
 		return nil, fmt.Errorf("Subscription must include the client's auth value")
 	}
 
+	plaintext := []byte(message)
+	var maxPayloadLength = aes128gcmMaxPayloadLength
```
Comment (x1ddos) [inline webpush/encrypt.go]: `maxPayloadLength := aes128gcmMaxPayloadLength`, also more consistent with the line above it.

File: webpush/encrypt.go [hunk 26]
```diff
@@ -55,13 +55,43 @@ import (
...
+// ContentEncoding indicates the version of encoding.
+type ContentEncoding int
+
 const (
-	maxPayloadLength = 4078
+	// The encoding that is widely deployed with WebPush (as of 2016-11).
+	AESGCM ContentEncoding = iota
+	// The most recent encoding, the salt, record size and key identifier
+	// are included in a header that is part of the encrypted content coding.
+	AES128GCM
```
Comment (x1ddos) [inline webpush/encrypt.go:66]: I wonder if `GCM` part is actually needed at all? Could it be just `AES` and `AES128`? I mean, is there any other `AESXXX` possible? Dropping GCM everywhere would make things much easier to read and use, imho.

Here and for all other variables/consts/funcs.

File: webpush/encrypt.go [hunk 27]
```diff
@@ -238,14 +293,48 @@ func newContext(clientPublicKey, serverPublicKey []byte) []byte {
 // The context argument should match what newContext creates
 func newInfo(infoType string, context []byte) []byte {
 	var info []byte
-	info = append(info, []byte("Content-Encoding: ")...)
+	info = append([]byte{}, []byte("Content-Encoding: ")...)
```
Comment (x1ddos) [inline webpush/encrypt.go]: nit: no need to define `var info []byte` really. I personally find it easier to read like this:

```go
info := append([]byte("Content-Encoding: "), []byte(infoType)...)
info = append(info, 0)
```

File: webpush/encrypt.go [hunk 28]
```diff
@@ -144,20 +174,27 @@ type EncryptionResult struct {
...
 	}
 
+	plaintext := []byte(message)
+	var maxPayloadLength = aes128gcmMaxPayloadLength
+	if encoding == AESGCM {
+		maxPayloadLength = aesgcmMaxPayloadLength
+	}
+
+	if len(plaintext) > maxPayloadLength {
+		return nil, fmt.Errorf("Payload is too large. The max number of bytes is %d, input is %d bytes.", maxPayloadLength, len(plaintext))
```
Comment (x1ddos) [inline webpush/encrypt.go]: nit, if you want, use a variable:

```go
if n := len(plaintext); n > maxPayloadLength {
  return nil, fmt.Errorf("...", maxPayloadLength, n)
}
```
so you don't have to `len(plaintext)` twice.

File: webpush/encrypt.go [hunk 29]
```diff
@@ -238,14 +293,48 @@ func newContext(clientPublicKey, serverPublicKey []byte) []byte {
...
 
+// Returns an 86 octet header. See section 2.1 of:
+// https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-06#section-2.1
+func appendHeader(salt, serverPublicKey, ciphertext []byte) []byte {
+	var result []byte
+	// A push service is not required to support more than 4096 octets of
+	// payload body so the record size can be at most 4096.
+	rs := uint32(maxPayloadRecordSize)
+	rsbuf := make([]byte, 4)
+	binary.BigEndian.PutUint32(rsbuf, rs)
```
Comment (x1ddos) [inline webpush/encrypt.go]: no need for `rs` really, just do `PutUint32(rsbuf, uint32(maxPayloadRecordSize))`

File: webpush/encrypt_test.go [hunk 30]
```diff
@@ -202,6 +231,37 @@ func TestRfcVectors(t *testing.T) {
...
+	}
+	key, err := b64.DecodeString(rfcAES128gcmPublic)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sub := &Subscription{Auth: auth, Key: key}
+	result, err := Encrypt(sub, aes128gcmMessage, AES128GCM)
+	if err != nil {
+		t.Error(err)
```
Comment (x1ddos) [inline webpush/encrypt_test.go]: `t.Fatal` here and everywhere else, where the return value is being used. You can't rely on the `result` below if err is returned.

Comment (x1ddos) [reply webpush/encrypt_test.go]: Both, this test and `TestAESgcmRfcVectors`.

**Review by x1ddos — CHANGES_REQUESTED** [review]
Once last request: could you rebase this onto master, please. 
I don't see Travis CI results, plus there's a merge conflict anyway.

File: webpush/encrypt.go (hunk 26 above)
Comment (pinarx) [reply webpush/encrypt.go:66]: specs refer to them as aesgcm and aes128gcm so I'm leaning towards keeping it, but also don't feel strongly about it. So if you do feel strongly I can change it in the next round.

File: webpush/encrypt.go (hunk 24 above)
Comment (pinarx) [reply webpush/encrypt.go]: reverted back to original in that case and returning unknown. Also added a check in newCEK and return an error if the encoding is unknown, but it's a bit awkward because they probably can't pass any value other than AES and AES128. Anyway take a look.

File: webpush/encrypt.go (hunk 25 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 28 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go (hunk 27 above)
Comment (pinarx) [reply webpush/encrypt.go]: ah agreed, thanks

File: webpush/encrypt.go (hunk 29 above)
Comment (pinarx) [reply webpush/encrypt.go]: done

File: webpush/encrypt.go [hunk 31]
```diff
@@ -237,15 +299,45 @@ func newContext(clientPublicKey, serverPublicKey []byte) []byte {
...
 // The context argument should match what newContext creates
 func newInfo(infoType string, context []byte) []byte {
-	var info []byte
-	info = append(info, []byte("Content-Encoding: ")...)
-	info = append(info, []byte(infoType)...)
-	info = append(info, 0)
-	info = append(info, []byte("P-256")...)
-	info = append(info, context...)
+	info := append([]byte("Content-Encoding: "), []byte(infoType)...)
+  info = append(info, 0)
```
Comment (x1ddos) [inline webpush/encrypt.go]: somehow this lost indentation? you should probably setup auto-gofmt on save :)

File: webpush/encrypt.go (hunk 24 above)
Comment (x1ddos) [reply webpush/encrypt.go]: Well, you can always pass something else if the base type is exported or built-in, like in this case.
Here's an example: https://play.golang.org/p/m-ffT2LvBH

You'd have to make ContentEncoding from some unexported type, but I don't think it's really worth it.

**Review by x1ddos — APPROVED** [review]

**Comment by x1ddos** [comment]
Just an indentation nit. Probably needs to be feeded to gofmt again. It's a good practice run it every time on save.

@gauntface I wonder where Travis CI results though. Can't see them here in the PR.

File: webpush/encrypt.go (hunk 31 above)
Comment (pinarx) [reply webpush/encrypt.go]: good idea

Comment (pinarx) [reply webpush/encrypt.go]: done

**Comment by x1ddos** [comment]
Ah, @gauntface I know why. You still need to submit your #22.
